brave/src/main/java/brave/Tracing.java

```diff
@@ -393,34 +393,30 @@ public Tracing build() {
       this.sampler = builder.sampler;
       this.noop = new AtomicBoolean();
 
-      ArrayList<FinishedSpanHandler> finishedSpanHandlers =
-        new ArrayList<>(builder.finishedSpanHandlers); // defensive copy against mutation of builder
+      FinishedSpanHandler zipkinHandler = builder.spanReporter != Reporter.NOOP
+        ? new ZipkinFinishedSpanHandler(builder.spanReporter, errorParser,
+        builder.localServiceName, builder.localIp, builder.localPort)
+        : FinishedSpanHandler.NOOP;
 
-      // If a Zipkin reporter is present, it is invoked after the user-supplied finished span handlers.
-      FinishedSpanHandler zipkinHandler = FinishedSpanHandler.NOOP;
-      if (builder.spanReporter != Reporter.NOOP) {
-        zipkinHandler = new ZipkinFinishedSpanHandler(builder.spanReporter, errorParser,
-          builder.localServiceName, builder.localIp, builder.localPort);
-        finishedSpanHandlers.add(zipkinHandler);
+      FinishedSpanHandler finishedSpanHandler =
+        zipkinReportingFinishedSpanHandler(builder.finishedSpanHandlers, zipkinHandler, noop);
+
+      ArrayList<FinishedSpanHandler> orphanedSpanHandlers = new ArrayList<>();
+      for (FinishedSpanHandler handler : builder.finishedSpanHandlers) {
+        if (handler.supportsOrphans()) orphanedSpanHandlers.add(handler);
       }
 
-      FinishedSpanHandler finishedSpanHandler;
-      switch (finishedSpanHandlers.size()) {
-        case 0:
-          finishedSpanHandler = FinishedSpanHandler.NOOP;
-          break;
-        case 1:
-          finishedSpanHandler = finishedSpanHandlers.iterator().next();
-          break;
-        default:
-          finishedSpanHandler = FinishedSpanHandlers.compose(finishedSpanHandlers);
+      FinishedSpanHandler orphanedSpanHandler = finishedSpanHandler;
+      if (!builder.finishedSpanHandlers.equals(orphanedSpanHandlers)) {
```

Comment:
It took a double-take (literally a look in a second review pass) to properly understand this comparison for some reason. Maybe clearer to set a boolean flag `notAllHandlersSupportOrphans` or something in the above loop instead.